### PR TITLE
feat: add collection holder tab

### LINKED ISSUE #2707: Add tab named Holders to collections to show owned assets by particular holders
We would like to see in the collection that the owner has which NFTs.

It would work we can list owners as a table and on click, it would show custom detailed rows
- headers would be 
  - owner / nft name
  - Bought - price bought for (BUY to user)
  - Sale - price selling for (LIST), 
  - Date (relative with a tooltip with exact one)
- link NFT name to particular nft item detail 
- link date to transaction same as in history table 

Use https://buefy.org/documentation/table/#custom-detailed-rows

Might need to write a resolver for this? 

### Ref
- #2222 
- #(collection by holders) something in passion feed iirc

### PR DESCRIPTION
**Pull request by Jarsen136**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### What's new?

- [x] PR closes #2707
- [ ] <brief_description_of_what_I've_added>

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

### Optional

- [x] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [x] I've tested PR on mobile and everything seems works
- [x] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)

### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.
image

### CODE REVIEW

**Comment by Jarsen136** [comment]
In addition, I added functions such pagination、Keyboard short cut and url page index.

**Review by roiLeo — CHANGES_REQUESTED** [review]
Design looks fine

Tested it on `/rmrk/collection/800f8a914281765a7d-KITTY?tab=holder` and I found some weird stuff:
- sorting only work on displayed data (there is an option with buefy to fix that with [`backend-sorting`](https://buefy.org/documentation/table/#sorting-multiple)
- when I'm on page 11, I see [this nft](https://deploy-preview-2724--koda-nuxt.netlify.app/rmrk/gallery/9068996-800f8a914281765a7d-KITTY-KITTY_PARADISE_16-0000000000000016) sale at 735.0000 KSM and when I check the nft to verify if the sale happened I see that the Item is not sale but listed at 735 ksm
- I skimmed the code and I have the impression that you are reusing code from `History` component, it would be super cool if you find a way to refactor the event loop

**Comment by yangwao** [comment]
quick thoughts 
- I would sort them default by the amount of NFTs address hold, add one more column, "assets" (?)  
- sort from bigger holders per address to lowest 

image

**Comment by Jarsen136** [comment]
> * sorting only work on displayed data (there is an option with buefy to fix that with [`backend-sorting`](https://buefy.org/documentation/table/#sorting-multiple))

Cool idea! I will have a try.

> * when I'm on page 11, I see [this nft](https://deploy-preview-2724--koda-nuxt.netlify.app/rmrk/gallery/9068996-800f8a914281765a7d-KITTY-KITTY_PARADISE_16-0000000000000016) sale at 735.0000 KSM and when I check the nft to verify if the sale happened I see that the Item is not sale but listed at 735 ksm

This is how it should behave, according to the description:
Bought - price bought for (BUY to user)
Sale - price selling for (LIST),
ref: https://github.com/kodadot/nft-gallery/issues/2707#issue-1184798294

> * I skimmed the code and I have the impression that you are reusing code from `History` component, it would be super cool if you find a way to refactor the event for loop

The event loop in `holder` is quite different to that in `history`. As you can see, very complex event loop logic is written to handle different cases. Maybe in the future, these list result can be provided by the backend, though it has a huge amount of work. : )

**Comment by Jarsen136** [comment]
> 

No problems, I will take it.

**Comment by Jarsen136** [comment]
I think it is better now, except placing the pagination at the bottom : )   @yangwao  @roiLeo 
I use the pagination provided by 'buefy' so that we can sort all the data and keep pagination.

image

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+            cell-class="short-identity__table"
+            :visible="columnsVisible['Bought'].display"
+            field="Bought"
+            label="Bought"
+            sortable
+            v-slot="props">
+            {{ props.row.BoughtFormatted }}
+          </b-table-column>
+          <b-table-column
+            cell-class="short-identity__table"
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:88]:
```suggestion
```

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+            v-slot="props">
+            <span v-if="props.row.Holder === '-'">{{ props.row.Holder }}</span>
+            <nuxt-link
+              v-else
+              :to="{ name: 'rmrk-u-id', params: { id: props.row.Holder } }">
+              <Identity :address="props.row.Holder" inline noOverflow />
+            </nuxt-link>
+          </b-table-column>
+          <b-table-column
+            cell-class="short-identity__table"
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:70]:
```suggestion
            numeric
```

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+      :class="hideCollapse ? 'collapseHidden' : 'bordered'"
+      animation="slide"
+      aria-id="contentIdForHistory">
+      <template #trigger="props">
+        <div
+          class="card-header"
+          role="button"
+          aria-controls="contentIdForHistory">
+          <p class="card-header-title">
+            {{ $t('History') }}
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:15]:
```suggestion
            {{ $t('Holders') }}
```

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+            cell-class="short-identity__table"
+            :visible="columnsVisible['Amount'].display"
+            field="Amount"
+            label="Amount"
+            sortable
+            v-slot="props">
+            {{ props.row.Amount }}
+          </b-table-column>
+          <b-table-column
+            cell-class="short-identity__table"
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:79]:
```suggestion
```

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+            :visible="columnsVisible['Sale'].display"
+            field="Sale"
+            label="Sale"
+            sortable
+            v-slot="props">
+            {{ props.row.SaleFormatted }}
+          </b-table-column>
+          <b-table-column
+            cell-class="short-identity__table"
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:97]:
```suggestion
```

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+                  :to="{
+                    name: 'rmrk-gallery-id',
+                    params: { id: item.Item.id },
+                  }">
+                  {{ item.Item.name }}
+                </nuxt-link>
+              </td>
+              <td
+                v-show="columnsVisible['Amount'].display"
+                class="short-identity__table">
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:129]:
```suggestion
                class="has-text-right">
```

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+                {{ item.Amount }}
+              </td>
+              <td
+                v-show="columnsVisible['Bought'].display"
+                class="short-identity__table">
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:134]:
```suggestion
              <td v-show="columnsVisible['Bought'].display">
```

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+                {{ item.BoughtFormatted }}
+              </td>
+              <td
+                v-show="columnsVisible['Sale'].display"
+                class="short-identity__table">
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:139]:
```suggestion
              <td v-show="columnsVisible['Sale'].display">
```

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+                {{ item.SaleFormatted }}
+              </td>
+              <td
+                v-show="columnsVisible['Date'].display"
+                class="short-identity__table">
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:144]:
```suggestion
              <td v-show="columnsVisible['Date'].display">
```

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+      .catch(console.warn /*Navigation Duplicate err fix later */)
+  }
+}
+</script>
+<style lang="scss">
+.short-identity__table {
+  max-width: 50em;
+  overflow: hidden;
+  text-overflow: ellipsis;
+}
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:416]:
```suggestion
```
this already defined in History component

File: components/rmrk/Gallery/Holder/Holder.vue [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+  .table-nav {
+    display: flex;
+    justify-content: space-between;
+  }
+  .box-container {
+    @media screen and (max-width: 768px) {
+      flex-direction: column-reverse;
+    }
+  }
+}
```
Comment (roiLeo) [inline components/rmrk/Gallery/Holder/Holder.vue:438]: this will be global style, could you try to make it scoped?

**Review by roiLeo — COMMENTED** [review]
some slight changes to take into account

File: components/rmrk/Gallery/Holder/Holder.vue (hunk 11 above)
Comment (Jarsen136) [reply components/rmrk/Gallery/Holder/Holder.vue]: ok

**Review by roiLeo — COMMENTED** [review]
looks better, I found that Date sorting doesn't work as excepted, it it's not possible I would suggest to remove the sortable option

For nft holden by an adress, is it possible to sort them by last bought (date)?
Screenshot 2022-04-01 at 17-02-53 Donkey Gang

**Comment by Jarsen136** [comment]
> looks better, I found that Date sorting doesn't work as excepted, it it's not possible I would suggest to remove the sortable option
> 
> For nft holden by an adress, is it possible to sort them by last bought (date)?
> 


It seem that it's impossible to sort the nfts by an address, because the `buefy ` does not provide that sorting inside a group.  

image

**Review by roiLeo — APPROVED** [review]
looks good, we can add the sorting by date in the next iteration

**Comment by yangwao** [comment]
Implementing sorting by date under address could be interesting, I'll make this as follow up tho!

**Comment by yangwao** [comment]
pay 300 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $300 @ 193.82 USD/KSM ~ 1.548 $KSM
🧗 [Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG](https://kusama.subscan.io/account/Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)
🔗 [0x1ab8e186020db6d20ab46273c1fef380014832b0890e8fb04ea6a85c189266ca](https://kusama.subscan.io/extrinsic/0x1ab8e186020db6d20ab46273c1fef380014832b0890e8fb04ea6a85c189266ca)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**